File: chae-dahee/ch05/2644.js
let input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const n = Number(input.shift()); // 전체 사람 수
const [a, b] = input.shift().split(' ').map(Number); // a, b 두 사람의 번호
const m = Number(input.shift()); // 관계의 개수
const arr = input.map((v) => v.split(' ').map(Number)); // 부모-자식 관계 배열

let visited = Array(n + 1).fill(false);
let graph = [...Array(n + 1)].map(() => []);

// 양방향 그래프 만들기
arr.map(([from, to]) => {
  graph[from].push(to);
  graph[to].push(from);
});

// DFS
const dfs = (start, target) => {
  // stack 초기값 설정
  let stack = [[start, 0]];

  // 방문 처리
  visited[start] = true;

  while (stack.length) {
    const [curNum, depth] = stack.pop();

    if (curNum === target) return depth;

    for (const node of graph[curNum]) {
      if (!visited[node]) {
        visited[node] = true;
        stack.push([node, depth + 1]);
      }
    }
  }
  return -1;
};

console.log(dfs(a, b));

File: jooyeong/ch13/2644.py
# DFS

import sys
input = sys.stdin.readline

n = int(input())
c1, c2 = map(int, input().split())
t = int(input())
d = [-1] * (n + 1)

g = [[] for _ in range(n + 1)]
for _ in range(t):
    p, c = map(int, input().split())
    g[p].append(c)
    g[c].append(p)

def dfs(g, w):
    for i in g[w]:
        if i == c2:
            d[i] = d[w] + 1
        if d[i] == -1:
            d[i] = d[w] + 1
            dfs(g, i)

d[c1] = 0
dfs(g, c1)